Create the db schema with Sequelize

TimHenkelmann opened the pull request:
The "create_sequelize_schema.js" file will populate the database defined in the config.js file with the following entities:
- artist
- release
- recording
- work
- group
- instrument

TimHenkelmann committed:
Create the db schema with Sequelize

The "create_sequelize_schema.js" file will populate the database defined in the config.js file with the following entities:
- artist
- release
- recording
- work
- group
- instrument

felixschorer commented:
Great! @ShilpaGhanashyamGore also mentioned a table with all entities? We also need 2 additional tables. One to store website metadata and one to store mappings from entities to websites. See `Desired Table Structure` in Google Docs.

felixschorer commented:
The sequelize code should ideally be in the `api/helpers` folder

TimHenkelmann commented:
Alright, then I'll move them there and create a separate sequelize file for your tables!

TimHenkelmann committed:
Moved "create_sequelize_schema" files to /api/helpers/

kordianbruck commented on api/helpers/create_sequelize_schema/config.js:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Created by TimH on 16-Mar-17.
+ */
```
Please remove those kinds of comments.

TimHenkelmann replied:
Whoops, sorry!

kordianbruck commented on api/helpers/create_sequelize_schema/config.js:
```diff
@@ -0,0 +1,12 @@
+config.port = "5432";
+config.database = "postgres";
+config.username = "postgres";
+config.password = "1234";
```
Please make this file `json` so that you don't need to add exports and no one will ever dream of adding logic here.

kordianbruck commented on api/helpers/create_sequelize_schema/create_sequelize_schema.js:
```diff
@@ -0,0 +1,158 @@
+/**
+ * Created by TimH on 16-Mar-17.
+ */
```
again.

kordianbruck commented on api/helpers/create_sequelize_schema/package.json:
```diff
@@ -0,0 +1,17 @@
+  "dependencies": {
+    "pg": "^6.1.4",
+    "pg-hstore": "^2.3.2",
+    "sequelize": "^3.30.2"
```
Please lock down the versions (remove ~ or ^) or use yarn (https://yarnpkg.com)

TimHenkelmann committed:
Applied changes:
- Removed "create" comments
- changed config to .json
- locked down versions in package.json

TimHenkelmann commented:
@kordianbruck applied your changes!

felixschorer commented:
@TimHenkelmann would it also be possible to add the two tables requested in #10  so we have a way to store our metadata? According to @ShilpaGhanashyamGore you already have a table with all entities?

sacdallago commented:
I like this!

sacdallago reviewed (changes requested):
I like this! But be aware that you will need the [Sequelize _classes_](https://github.com/MusicConnectionMachine/api/pull/11/files#diff-a0babe3ac085379280c3175db75ca2b9R24) throughout the API in order to make full use of the ORM idea.

Also. Please be a bit more reusable.

What you can do is the following:
1. Create a folder [here](https://github.com/MusicConnectionMachine/api/tree/develop/api) called `models`
2. For every model in [this](https://github.com/MusicConnectionMachine/api/pull/11/files#diff-a0babe3ac085379280c3175db75ca2b9R1) file, create a separate file, construct the table and export the class (e.g.`artist` --> `artist.js` --> `module.exports = new Sequelize.define('artists')`, `release`, ...).
3. Import and use the definitions where they are needed (E.g. In the controllers).
4. Import all definitions on bootstrap (eg. somwhere [here](https://github.com/MusicConnectionMachine/api/blob/develop/app.js#L11 ) and `sync` the tables).

sacdallago commented:
[UPDATE]
@TimHenkelmann I've added you to a private repo of mine, if you navigate to [`app/models`](https://github.com/sacdallago/mecu/tree/master/app/models) you see what I mean. Also interesting are [this](https://github.com/sacdallago/mecu/blob/master/index.js#L56) and [this](https://github.com/sacdallago/mecu/blob/master/app.js#L58)

💃

TimHenkelmann commented:
@lgfrbcsgo sorry for the late reply. Yes, I will also add those tables requested in #10. But as stated in #9 , the "entities" table is dropped.

TimHenkelmann commented:
@lgfrbcsgo I just saw that you will need the "entities" table for the "contains" table as a foreign key as stated in #10 , therefore we cannot remove it. Damn :D 
So my suggestions would be to create an "entities" table which will have the columns:
entity_id, entity_type, type_id
where type can be "artist", "work", "release" or "instrument" and type_id will have the id in the respective table. This will allow us to reference all entities from one table without the issue of handling duplicate data (once in the entity table, once in the artist/work/release/instrument table). 
If you approve that, I will add this table to the Sequelize scripts

TimHenkelmann committed:
Moved sequelize.defines into separate models folder
- still an issue with: password authentication failed for user "musicmachine"

TimHenkelmann commented:
Currently it fails with the error: "Unhandled rejection SequelizeConnectionError: password authentication failed for user "musicmachine" => needs research

lgfrbcsgo commented:
@TimHenkelmann sound great!

felixschorer commented:
@TimHenkelmann sounds great! Or you could add an extra column to all other tables containing a foreign key to entities. That way everything stays coherent when deleting stuff.

```
Entities(
    id:         int, primary key,
    type:       varchar
)

Artists(
    id:         int, primary key,
            ...
    entityId:   int, foreign key -> Entities
)
```

felixschorer commented:
You have to use the keyword `on delete cascade` though

kordianbruck commented:
@sacdallago can you update your review please so we can merge?

TimHenkelmann committed:
Fixed config+sync issue

TimHenkelmann committed:
made table names plural

TimHenkelmann commented:
Fixed some issues and renamed table names to plural. Ready for review!

ShilpaGhanashyamGore reviewed (approved):
Looks good!

TimHenkelmann committed:
Moved sequelize + table initialization into app.js
Moved models folder from /api/helpers to /api

TimHenkelmann commented:
@sacdallago I just moved all the definitions and sequelize.sync into app.js and moved the models folder to /api. Could you please review it and delete your change request if everything looks good?

kordianbruck commented:
@TimHenkelmann he seems not to be around. I've merged it for you - looks good. Chris can still add his remarks and you'll just add a new PR then.
